substrate/client/network/src/litep2p/shim/notification/peerset.rs

```diff
@@ -0,0 +1,1276 @@
+										log::debug!(target: LOG_TARGET, "{}: {peer:?} doesn't exist", self.protocol);
+										debug_assert!(false);
+										None
+									},
+								}
+							})
+							.collect(),
+					}))
+				},
+				PeersetCommand::AddReservedPeers { peers } => {
+					log::debug!(target: LOG_TARGET, "{}: add reserved peers {peers:?}", self.protocol);
+
+					for peer in &peers {
+						let mut num_out = 0;
+						let mut num_in = 0;
+
+						match self.peers.get_mut(&peer) {
+							Some(PeerState::Disconnected | PeerState::Backoff) => {},
+							Some(
+								PeerState::Opening { ref mut direction } |
+								PeerState::Connected { ref mut direction } |
+								PeerState::Canceled { ref mut direction } |
+								PeerState::Closing { ref mut direction },
+							) => {
+								*direction = match direction {
+									Direction::Inbound(Reserved::No) => {
+										num_in += 1;
+										Direction::Inbound(Reserved::Yes)
+									},
+									Direction::Outbound(Reserved::No) => {
+										num_out += 1;
+										Direction::Outbound(Reserved::Yes)
+									},
+									ref direction => **direction,
+								};
+							},
+							None => {
+								self.peers.insert(*peer, PeerState::Disconnected);
+							},
+						}
+
+						self.num_in -= num_in;
+						self.num_out -= num_out;
+					}
+
+					let peers = peers
+						.iter()
+						.filter_map(|peer| {
+							if !self.reserved_peers.insert(*peer) {
+								log::warn!(
+									target: LOG_TARGET,
+									"{}: {peer:?} is already a reserved peer",
+									self.protocol,
+								);
+								return None
+							}
+
+							std::matches!(
+								self.peers.get_mut(peer),
+								None | Some(PeerState::Disconnected)
+							)
+							.then(|| {
+								self.peers.insert(
+									*peer,
+									PeerState::Opening {
+										direction: Direction::Outbound(Reserved::Yes),
+									},
+								);
+								*peer
+							})
+						})
+						.collect();
+
+					log::debug!(target: LOG_TARGET, "{}: start connecting to {peers:?}", self.protocol);
+
+					return Poll::Ready(Some(PeersetNotificationCommand::OpenSubstream { peers }))
+				},
+				PeersetCommand::RemoveReservedPeers { peers } => {
+					log::debug!(target: LOG_TARGET, "{}: remove reserved peers {peers:?}", self.protocol);
+
+					return Poll::Ready(Some(PeersetNotificationCommand::CloseSubstream {
+						peers: peers
+							.iter()
+							.filter_map(|peer| {
+								if !self.reserved_peers.remove(peer) {
+									log::debug!(
+										target: LOG_TARGET,
+										"{}: {peer} is not a reserved peer",
+										self.protocol,
+									);
+									return None
+								}
+
+								let protocol = self.protocol.clone();
+								let peer_state = self.peers.get_mut(peer)?;
+
+								match std::mem::replace(peer_state, PeerState::Disconnected) {
+									PeerState::Connected { direction } => {
+										*peer_state = PeerState::Closing { direction };
+										return Some(*peer)
+									},
```

Comment:
Shouldn't we try moving the peer into the regular peers set, without disconnecting it in case we have the slots available? At least it was like this in the original implementation & `ProtocolController`.

Reply:
Yes we should, I'll implement it

Reply:
Done